---
title: "REAME: SARS-CoV-2 sequence submission process"
author:
- Mandy Waters
- CDPHE Contractor
upadated by:
- Molly Hetherington-Rauth, Dec 6, 2021
date: "July 2021"
output:
  html_document:
    df_print: paged
    highlight: tango
    theme: flatly
fontsize: 12pt
---

```{r "load_packages", include=FALSE, message=FALSE}
  library(knitr)
  library(kableExtra)
  library(png)
  library(grid)
```

###Goals
Provide a detailed overview of SARS-CoV-2 submissions to GISAID, NCBI BioSample, NCBI SRA, and NCBI GenBank. This document will take CDPHE colleagues through initial set up on a GCP VM running Ubuntu 18.04, process / pipeline flow, and scripts developed to automate metadata genereation. Scripts are availble on the CDPHE [Github](https://github.com/CDPHE/seq-submission-metadata-R). This section will cover the following items:

1. Getting set up with software and accounts (GISAID and NCBI)  
2. High level overview and process  
3. General information

###Set up and software {.tabset .tabset-fade .tabset-pills}
####Overview
Users will need to create accounts with GISAID and NCBI, install software, and complete some prliminary steps before running the sequence submission pipeline.  Please follow all instructions in subsequent tabs to get set up.

####GISAID Account
Navigate to GISAID's registration page: [link](https://www.gisaid.org/registration/register/).  
Click on the 'register' button on the bottom right corner of the page.  Fill in with your personal and CDPHE info:  

Institution: Colorado Department of Public Health and Environment  
Department: Microbiology   
Address: 8100 E Lowry Blvd Denver CO 80230  
email: CDPHE email  

They will send you an email to confirm your registration with a link you need to click on. Then you will get another email confirming your successful registration.

####NCBI Account
NCBI has a new login method where you use your gmail (CDPHE email) to sign in and authenticate. Make sure the keychain for your browser is for your CDPHE email otherwise it will error out. I had some intermitent issues getting proxy errors and such, but kept refreshing until it worked. Follow the instructions given by NCBI to create a new account. NCBI will send you a confirmation email once they've processed your request.  

Navigate to NCBI's registration page: [link](https://account.ncbi.nlm.nih.gov/signup/).  

Institution: Colorado Department of Public Health and Environment  
Department: Microbiology   
Address: 8100 E Lowry Blvd Denver CO 80230  
email: CDPHE email  
back-up email: personal email

####Software

Programs to download/install on your VM for the submission process:  
1. [Atom](https://flight-manual.atom.io/getting-started/sections/installing-atom/) - or your other favorite Linux Text Editor    
2. [Seqtk](https://github.com/lh3/seqtk): parsing and editing fasta files.  Link takes you to a github repo with a readme  
3. [SeqKit](https://bioinf.shenwei.me/seqkit/download/): parsing and editing fasta files  
4. [VADR](https://github.com/StaPH-B/staphb_toolkit): is part of the staphb-tk that you installed in the Getting Started document  
5. [MAFFT](https://mafft.cbrc.jp/alignment/software/linux.html): MSA.  Molly's version of indel_finder as of Sept 3, 2021 runs MAFFT pairwise for each sample in your batch, but might be good to have a local copy   
6. [IGV](https://software.broadinstitute.org/software/igv/download): checking read coverage  
7. [Indel_finder.py](https://github.com/CDPHE/sars-cov-2_indel_finder): is on the CDPHE github page  
8. Samtools: working and indexing alignment (bam) files

```{bash, "install_samtools", echo=TRUE, eval=FALSE}
# I have conda installed and therefore was able to install samtools with conda
conda install -c bioconda samtools
```

9. Aspera Connect Ascp: gsutil cp -r gs://laura-sandbox/aspera-connect/ . Then you can find ascp in aspera/connect/bin.  You might need to make the ascp file executable by running chmod +x aspera/connect/bin/ascp
10. [Aliview](https://ormbunkar.se/aliview/#DOWNLOAD) - review MSA.  After install you can open 'aliview' by just typing aliview on your VM terminal
11. rename_fasta.py: might be good to have, but Molly's modifications to indel_finder as of Sept 3, 2021 takes in multi-fasta, so this script isn't needed.  Can be accessed using: gsutil cp gs://laura-terra/rename_fasta.py .  

###Overview of process
Below is a flowchart showing the major steps in submitting SARS-CoV-2 samples to GISAID and NCBI.  A ppt version of this diagram can be found the CDPHE [Github](https://github.com/CDPHE/seq-submission-metadata-R) as seq_upload_flowchart.pptx

```{r, "overview_flowchart", fig.width=8, fig.height=5, fig.align = "center", echo=FALSE, warning=FALSE}
img <- readPNG("/Users/water157/Desktop/OneDrive - Pepsico/covid/seq_sub_overview.png")
grid.raster(img)
```

###General information {.tabset .tabset-fade .tabset-pills}
####Finding data
Special call out to Laura Bankers for most of the content below

**Overview:** All sequencing data, intermediate files, alignments (bam files), and assemblies are stored in the google bucket gs://covid_terra in the SNP Linux project on GCP. Inside that bucket, there is a directory for each sequencing run, and inside that directory are all of the raw data and necessary files for every sample sequenced on that run. Sequencing run directories are named based on the platform + run id number. Those beginning with COVSEQ_ are paired-end Illumina Miseq runs, those beginning with COVMIN_ are Oxford Nanopore GridION runs, NEXSEQ_ are single-end NextSeq runs, and COVARC are archived samples from earlier in the pandemic that we are now going back to and are sequenced on either MiSeq or GridION. The numbers after the underscore are the run id numbers (e.g. COVMIN_0009 is nanopore run 0009).  For submissions you will need the consensus genome fasta files, the raw reads, and the metadata. If you need to go back and check the validity of some SNP or indels calls, you may also need to bam files  

**Metadata:** For all COVMIN and NEXSEQ runs, and all COVSEQ runs from COVSEQ_0062rq and greater, metadata is found in the results_RUNNAME_RUNDATE file e.g. (results_COVMIN_0009_2021-04-27), which are pushed and can be downloaded from a shared Google Drive

**Consensus genome fasta files:** the final consensus genome fasta files to be submitted are named accessionid_renamed_consensus.fa (e.g. 2100213088_consensus_renamed.fa) and is located in gs://covid_terra/run_directory/terra_outputs/assemblies/ (e.g. gs://covid_terra/COVMIN_0009/terra_outputs/assemblies/)  

**Raw reads:** For Illumina MiSeq and NextSeq runs (COVSEQ and NEXSEQ), raw reads for each sample on the run can be found in gs://covid_terra/COVSEQ_NNNN/fastq_files/. For nanopore runs (COVMIN), the fastq files for each sample are in separate directories that are named based on each samples barcode (e.g. gs:// covid_terra/COVMIN_0009/fastq_pass/barcode02), which need to be merged together (script built to do this)  

**Bam files:** bam files can be found in gs:// covid_terra/SEQRUN_NNNN/terra_outputs/alignments (e.g. gs:// covid_terra/COVMIN_0009/terra_outputs/alignments)  

**Samples that have been re-run / resequenced:** Some samples needed to be resequenced. A list of these samples can be found [here](https://docs.google.com/spreadsheets/d/1PtaHHa8iuqM6fNGTSkHiinxc5poQnVt1Ti67yLSVQ6c/edit#gid=471638297).  This file will need to be downloaded before each submission so samples are not submitted more than once and the resequenced samples are being uploaded.

**Extra information:**

1. early COVMIN runs were re-analyzed due to workflow updates. The final results for these runs are in “terra_outputs_rr” (be sure to use files in the _rr directories in these cases). The underlying directory structure in the _rr directory is the same as the other output directories  
2. NCBI BioSample only lets you submit up to 1000 samples at a time – So I tend to just maintain batch numbers of up to 1000 sample throughout the entire process   
3. GenBank only allows you to select a single sequencing technology for an entire submission. So only include a single sequencing platform in a given batch (i.e. up to 1000 sample from only Illumina MiSeq runs)

####Tracking submissions
Laura has created an [excel document](https://docs.google.com/spreadsheets/d/1n9NzJNrqIHP9Lct_MZuSnOv3Qn1JCZsI/edit#gid=331506794) to track which projects / batches have not been submitted, are in process, and are completed as well as tabs that outlines metadata sheets necessary for GISAID, BioSample, SRA, and GenBank (R scripts make these).

Sumbitters can add their initals next to the projects they are working on in the in the 'Notes' column in 'NOT_submitted' tab to ensure mulitple submitters aren't working on the same projects.  While we are catching up we are submitting samples  we are having GenBank reject samples when they find errors.  Add these samples to the 'ONGOING_submissions' tab and we will go back when we are more caught up. Once the samples are submitted to all public databases then all concatted metadata are put into the 'COMPLETED_submissions' tab.

####Download re-run samples
The sequencing team adds information about which samples were rerun on the CDPHE Google Drive.  You can download the re-run tab from 'List_of_bad_runs_and_waste_water_samples' google sheet using the instructions below.

1. Navigate to the [Google Sheet](https://docs.google.com/spreadsheets/d/1PtaHHa8iuqM6fNGTSkHiinxc5poQnVt1Ti67yLSVQ6c/edit#gid=471638297) - if you can't reach this site then one of the current contributors with edit access to share the file with you.
2. Click File -> Download -> tab-separated values (tsv, current sheet)
3. Once downloaded run the following command to pseudo track versions - this adds the date and gets rid of spaces

```{bash, "rename_reruns", echo=TRUE, eval=FALSE}
mv List_of_bad_runs_and_waste_water_samples\ -\ re-run_samples.tsv rerun_samples_$(date +%d%b%Y).tsv
```

####Download completed list
Some samples will be sequenced more than once on different sequencing platform (didn't fail, so they won't be in the rerun / bad noodle samples).  To account for this users should also download a copy of the COMPLETED_submissions' tab in submissions_and_compiled_metadata

1. Navigate to the [Google sheet](https://docs.google.com/spreadsheets/d/1n9NzJNrqIHP9Lct_MZuSnOv3Qn1JCZsI/edit#gid=331506794) - if you can't reach this site then one of the current contributors with edit access to share the file with you.
2. Click File -> Download -> tab-separated values (tsv, current sheet)
3. Once downloaded run the following command to pseudo track versions - this adds the date and gets rid of spaces

```{bash, "rename_completed", echo=TRUE, eval=FALSE}
mv SUBMISSIONS_and_compiled_metadata.xlsx\ -\ COMPLETED_submissions_metadata.tsv completed_submissions_$(date +%d%b%Y).tsv
```

####Download on-going list
Some samples will be rejected by NCBI Genbank, but will be submitted to GISAID, BioSample, and SRA, so we will also want to delete these out of your current submission as GISAID and NCBI (BioSample and SRA) will reject them.  Users should download a copy of the 'ONGOING_submissions' tab in submissions_and_compiled_metadata

1. Navigate to the [Google sheet](https://docs.google.com/spreadsheets/d/1n9NzJNrqIHP9Lct_MZuSnOv3Qn1JCZsI/edit#gid=80251444) - if you can't reach this site then one of the current contributors with edit access to share the file with you.
2. Click File -> Download -> tab-separated values (tsv, current sheet)
3. Once downloaded run the following command to pseudo track versions - this adds the date and gets rid of spaces

```{bash, "rename_onging", echo=TRUE, eval=FALSE}
mv SUBMISSIONS_and_compiled_metadata.xlsx\ -\ ONGOING_submissions.tsv ongoing_submissions_$(date +%d%b%Y).tsv
```

####GISAID submission

Submit your sequence to GISAID

1. The metadata file needed will be written by the pytho script gisaid_metadata_formatting.R with all columns headers with missing data for any fields that we don’t fill out  
2. Login to your [GISAID account](https://www.gisaid.org/), then go to the EpiCoV tab, then select > batch upload (you can ignore the CLI notification window that will pop up because the standard submission method is super easy. Upload your metadata csv, upload your fasta, and select “check and submit"    
3. You will receive an email once your sequences have been processed with those that have been released and another if GISAID finds any errors. They will either ask you to email them back with coverage information or respond by accessing the 'unreleased' tab on your GISAID account (more recently the latter is more common). If the frameshift mutations don't match something that was already flagged by VADR, then check the bam file in IGV. Write out a new fasta file with the consensus sequences that were rejected (corrected and unchanged sequences if coverage supports the nucleotide seq) and replace this new fasta in the upload page on GISAID.  

    Note that the email you get once the data are processed will contain all the GISAID accessions, but it will not say what samples they are linked to, so to get that information as well as Pangolin/Nextclade lineage information, you will need to go back to your gisaid account > search > fill in the submission date, and then download sequence technology metadata file.

####BioSample submission info

Submit to NCBI BioSample
note: we have updated our workflow a bit here. now we will include the gisaid accession ids in the biosample metadata file so we will need to wait to get the gisaid accession ids before continuing on to any of the ncbi submissions.

1. In your gisaid account, go to the search tab and download the "Sequencing technology metadata" file. Move this file into the metadata directory.
2. Create the ncbi biosample metdata file by using the ncbi_biosample_metadata_formatting.py script. You will only need to specify the path to the metadata directory.

```ncbi_biosample_metadata_formatting.py -m {metdata_dir_path}
 ```
3. Check for warning. The script will notify you if any sample is missing a gisaid accession id. If it a sample is missing a gisaid accession id, your gisaid submisson may not be completed processing.   
4. Login to NCBI and go the submission portal: https://submit.ncbi.nlm.nih.gov/subs/
5. Go to My Submissions  > BioSample > New Submission  
6. Make sure you use BioProject PRJNA686984.
7. Fill out the submitter info with your state email as the primary email and fill out institutional information as follows and be sure to use the cdphe submission group:

```{r, "bioosample_flowchart", fig.width=8, fig.height=5, fig.align = "center", echo=FALSE, warning=FALSE}
img <- readPNG("/Users/water157/Desktop/OneDrive - Pepsico/covid/biosample_info.png")
grid.raster(img)
```
8. Under General Information select:  
    + “Release immediately following processing”  
    + “BatchMultiple BioSamples” **Note that you are only allowed to submit up to 1000 Biosamples at a time  
9. Under Sample Template select “SARS-CoV-2: clinical or host-associated”  
10. Under Attributes you will upload your metadata tsv file  
11. Under Review and Submit, double check that everything looks correct and then hit submit  

Once your BioSample submission has been fully processed, you can proceed to NCBI SRA submission (you will need for BioSample SAMNNNNNNNN accession ids for SRA) – this generally takes less than an hour.

####SRA submission info

Submitting to NCBI SRA

1. Download the raw fastq files from the google bucket to you VM.
    + There will be two fastq files for each sample for Illuimina MiSeq runs (which are paired-end), and _R1 and an R2 fastq file. These are located in gs:// covid_terra/COVSEQ_NNNN/fastq_files/ and are all named as “accesionid_R1_001.fastq.gz” and “accesionid_R1_002.fastq.gz”

Code to pull reads for COVSEQ
```{bash, "download_illumina", echo=TRUE, eval=FALSE}
# Pull the fastq files from the google bucket onto your VM
Add code after it is tested
```

 Code to pull reads for NEXSEQ
```{bash, "download_illumina_next", echo=TRUE, eval=FALSE}
# Pull the terra data tables for the sequencing runs
while read name
do
gsutil cp gs://covid_terra/${name}/${name}_terra_data_table.tsv .
done < ${cur_wd}/project_list.txt

# Pull seqs for SRA - in sequence_data dir
# WARNING: this code is specific for NEXSEQ
while read name
do
while read acc path
do
gsutil -m cp $path .
done < ${name}_terra_data_table.tsv
done < ${cur_wd}/project_list.txt
```

    There will be a single file for Illumina NextSeq runs (which are single-end). These are located in gs:// covid_terra/NEXSEQ_NNN/fastq_files/ are all named as “accesionid.fastq.gz”

Nanopore runs have a large number of fastq files associated with each sample. They are organized into barcode directories – one barcode directory per sample. E.g.: gs:// covid_terra/COVMIN_0009/fastq_pass/barcode02 contains all the fastq files for the sample that was barcoded with barcode02.

Use the script get_cat_gzip_fastq.py on [Github](https://github.com/CDPHE/seq-submission-metadata-R) to compile all the barcodes into a single sample. This will also gzip the files so the names match the output from the R/python scripts  
2. Prepare your metadata file using the R script: ncbi_sra_metadata_formatting.R
3. Go to your NCBI Submissions Portal > Sequence Read Archive > New Submission
4. Submitter info is filled out the same way as for the BioSample submission
5. Fill out General Information as in the image below (be sure to use BioProject PRJNA686984)

```{r, "sra_info", fig.width=8, fig.height=5, fig.align = "left", echo=FALSE, warning=FALSE}
img <- readPNG("/Users/water157/Desktop/OneDrive - Pepsico/covid/sra_gen_info.png")
grid.raster(img)
```

6. Under the SRA Metadata tab, upload your metadata tsv fromncbi_sra_metadata_formatting.R
7. Under Files, select “FTP or Aspera Command Line file preload”
    + Read the Aspera Commandline instructions tab, and download the key file to your VM
    + Go back to your VM commandline window, cd into the directory containing your fastq files and run:

```{bash, "sra_upload", echo=TRUE, eval=FALSE}
# SRA will provide the rest of subasp@upload.ncbi.nlm.nih.gov:upload on your upload page
/path/to/aspera-connect/connect/bin/ascp -i /path/to/aspera.openssh -d sequence_dir subasp@upload.ncbi.nlm.nih.gov:uploads/...
```

8. Once you see that the upload has completed in your command line window, go back to your submission and click the “select preload folder” button. Select the folder containing your fastq files, and then “use selected folder” (make sure you refresh until it lists the same number of files as you plant to submit).
9. Under Review & Submit, make sure that everything looks correct and then hit submit.
10. Once the submission is processed, you will receive an email – this can take a few days because the BioProject is set up such that all SRA submission go through NCBI’s human data scrubbing pipeline. Once this completes, you will receive the SRR accession ids that you need for Genbank

####GenBank submission info
1. Go to you NCBI Submissions portal > GenBank > New Submission  
2. Under Submission Type > What do your sequences contain? Selection “SARS-CoV2, Influenza…” then for Which virus? Also select “SARS-CoV-2.” You can leave the submission title blank, but I fill in with the batch number in case I need to quickly reference which set of samples were included  
3. Submitter Information is the same as for other NCBI submissions
4. Under sequencing technology select either “Illumina” for an batch of MiSeq or NextSeq samples, or “Other” if the batch is of assemblies from nanopore data (then put Oxford Nanopore in the field that appears). Then select assembled sequences
    + For Assembly inforamtion for Illumina MiSeq (COVSEQ runs prior to COVSEQ_0063) use minimap2 v.2.1, and iVar v.1.3.1
    + For Assembly information for Illumina data (all NEXSEQ, all NOVASEQ and COVSEQ runs begining with COVSEQ_0063) put BWA v.0.7.17-r1188, and iVar v.1.3.1
    + For Assembly information Nanopore assemblies put Arctic MinION JAN-2020
5. Under Sequences, select “release immediately following processing”
6. Then upload the fasta file you generated (GenBank doesn't accept dashes and you need to add the NCBI project number so GenBank adds it to the right project - code has been developed to do this as well)
7. A warning will pop up regarding N's and request to answer the question 'What do the internal NNN's represent?'.  Fill in with 'A region of estimated length between the sequenced regions based on an alignment to similar sequences or genome"' and click 'continue'
8. Sequence processing: 'During processing, should NCBI remove sequences with errors and process the rest?". Depending on where we are against historic data either select 'yes' - Genbank will remove these from your submission and release the rest.  The final Rscript will pull the metadata for the rejected samples to add to the 'ONGOING_submissions" tab".  If you select 'no' then you will go back and forth with GenBank with coverage information etc.
9. Select 'None of these' for naming convention
10. Under the “Source Modifiers” tab you will upload your GenBank metadata tsv file
11. Under References you will List: Laura Bankers, Molly C. Hetherington-Rauth, Diana Ir, Alexandria Rossheim, Michael A. Martin, Mandy Waters, Arianna Smith, Shannon R. Matzinger, Sarah Elizabeth Totten, Emily A. Travanty – If names are not correct, please change them in the gisaid template section of the submissions google sheet, so that they are up to date.
    + You have to manually put these in the first time and then you can select and apply this list for later submissions
12. Then select “Unpublished” and then continue (can leave the rest of the boxes blank)
13. Under Review and Submit, make sure everything looks correct, then submit

###Detailed instructions for running R code
The objective of this code is to automate the creation of metadata tables for GISAID, NCBI Biosample, NCBI SRA, NCBI Genbank, and data for CDPHE tracking (all public IDs linked).  This section of the document will outline the following items:

1. Installing R version 4 on GCP VM with Linux Ubuntu 18.04
2. Preparing the scripts and installing packages
3. Executing the R scripts
    + gisaid_ncbi_biosample_metadata_formatting.R
    + ncbi_sra_metadata_formatting.R
    + ncbi_genbank_metadata_formatting.R
    + submission_completed_metadata.R

I put all my metadata files (both those required as inputs and written out from these scripts) for a batch of uploads in a single directory, so the working directory is always the same.

###Installing R {.tabset .tabset-fade .tabset-pills}
####Overview
You will need at least R version 4 to load the packages used in these scripts. Instructions are outlined (in tabs) for users who will be installing R for the first time as well as those who have a previous version of R already installed. NOTE: running sudo apt update alone will not update R to the version needed.

####First R install
Click [here](https://www.digitalocean.com/community/tutorials/how-to-install-r-on-ubuntu-18-04-quickstart) to go to page with instuctions, which are also outlined below. The code below assumes you are running Ubuntu 18.04. You will need a different repository file if you are running a different version. This repository file will install regardless of your version and then will throw errors / won't update when you install a different version.

```{bash, "install_R", echo=TRUE, eval=FALSE}
# Step 1 — Add GPG Key
# Logged into your Ubuntu 18.04 server as a sudo non-root user, add the relevant GPG key.
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9

# Step 2 — Add the R Repository
sudo add-apt-repository 'deb https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/'
    # If you’re not using 18.04, find the relevant repository from the R Project Ubuntu list, named for each release.

# Step 3 — Update Package Lists
sudo apt update

# Step 4 — Install R
sudo apt install r-base
    # If prompted to confirm installation, press y to continue.

# Step 5 — Test Install
# Start R’s interactive shell as root.
sudo -i R
```

You should see something like:
```{r, "soln_overview", fig.width=5, fig.height=3, fig.align = "left", echo=FALSE, warning=FALSE}
img <- readPNG("/Users/water157/Desktop/OneDrive - Pepsico/covid/confirm_r_install.png")
grid.raster(img)
```

####Uninstall R and update
Installing base R on 18.04 will install R version 3.5.  If you have R < 4.0 you will need to uninstall R and reinstall R version 4.  [Here](https://genomespot.blogspot.com/2020/06/installing-r-40-on-ubuntu-1804.html) is a link to instructions to uninstall older versions of R and then install version 4 (same as previous tab). These instructions are also below.

Uninstalling old version of R
```{bash, "uninstall_R", echo=TRUE, eval=FALSE}
# Step 1 - Make sure you have these dependancies installed. If you don't, you're going to have trouble installing any R packages like R curl, tidyverse, devtools, etc
sudo apt install libssl-dev libcurl4-openssl-dev libxml2-dev

#Step 2 - Remove any existing R install.
sudo apt remove r-base* --purge

# Step 3 - Remove any remaining packages. It's better to install them "fresh". They will be in locations like /home/user/R/R-3.6.6 and /usr/lib/R/library/

# Step 4 - Remove any existing entry for R in the etc/apt/sources.list to install an older R version.
sudo nano /etc/apt/sources.list
# For example you might have something like this:
# deb https://cloud.r-project.org/bin/linux/ubuntu bionic-cran35/

```

Now you can install R version 4
```{bash, "install_R_2", echo=TRUE, eval=FALSE}
# Step 1 — Add GPG Key
# Logged into your Ubuntu 18.04 server as a sudo non-root user, add the relevant GPG key.
sudo apt-key adv --keyserver keyserver.ubuntu.com --recv-keys E298A3A825C0D65DFD57CBB651716619E084DAB9

# Step 2 — Add the R Repository
sudo add-apt-repository 'deb https://cloud.r-project.org/bin/linux/ubuntu bionic-cran40/'
    # If you’re not using 18.04, find the relevant repository from the R Project Ubuntu list, named for each release.

# Step 3 — Update Package Lists
sudo apt update

# Step 4 — Install R
sudo apt install r-base
    # If prompted to confirm installation, press y to continue.

# Step 5 — Test Install
# Start R’s interactive shell as root.
sudo -i R
```

You should see something like:
```{r, "soln_overview2", fig.width=5, fig.height=3, fig.align = "left", echo=FALSE, warning=FALSE}
img <- readPNG("/Users/water157/Desktop/OneDrive - Pepsico/covid/confirm_r_install.png")
grid.raster(img)
```

###Preparing script and installing packages {.tabset .tabset-fade .tabset-pills}
####Prep scripts
Outlined here is instructions to make the R script executable on your VM and load necessary packages.  First step is to clone the R scripts to your VM from the CDPHE Github [link](https://github.com/CDPHE/seq-submission-metadata-R).  

Now make the R scripts executable (for some reason R scripts do not default as executable)
```{bash, "chmod", echo=TRUE, eval=FALSE}
chmod +x gisaid_ncbi_biosample_metadata_formatting.R
```

Now install the packages as outlined in the 'Install packages' tab.

####Installing packages
A few different packages are needed to run these scripts.  The individual scripts will call the packages they need, but they must be installed on your VM first.

```{bash, "load_packages_vm", echo=TRUE, eval=FALSE}
# Open an R session from your VM terminal
R

# Within your R environment run the following commands
if (packageVersion("devtools") < 1.6) {
  install.packages("devtools")
}
devtools::install_github("hadley/lazyeval")
devtools::install_github("hadley/dplyr")
devtools::install_github("hadley/plyr")
devtools::install_github("r-lib/cpp11")
install.packages("tidyr")
install.packages('optparse')
install.packages('stringr')
install.packages('tibble')

# devtools took a couple mins to load, this is normal

# Close R
quit()
```
Now you're ready to run the script R scripts

###GISAID and NCBI Biosample {.tabset .tabset-fade .tabset-pills}
####Useage
In short, the script takes in user specified information to pull all sequencing metadata sheets within a directory (reads in all results_.\*.csv in directory), concats the  files, filters for samples > 50% coverage, and then creates / writes the formatted files for sample submission to GISAID and NCBI Biosample, a file that lists the current sample names, GISAID formatted names (users can then use awk command to rename samples), and NCBI GenBank formatted names (with BIOPROJECT and Surveillance data if available), a file if any samples are missing the collection date, a file if the samples have a collection date prior to 2020, a file if those samples have already been uploaded (resequenced, but not because they failed), and a file if samples have been rerun because they failed or had contamination. The metadata file will be written out for samples with >50 coverage, are the latest sequencing results, aren't a blank, and haven't been previously submitted (GISAID, BioSample, SRA, or GenBank).  You can run this R script directly from the command line.  You will need to supply 7 pieces of information as options / flags when running the script.

**Script name:** gisaid_ncbi_biosample_metadata_formatting.R  
**Input files:** results_\*.tsv (pull all files with this regex), rerun_samples_\*.tsv, completed_submissions_\*.tsv, ongoing_submissions\*.tsv
**Options / flags:** required to fill in all.  The code will check spelling on the sequencing information and will error out if something is misspelled (options in #4 below)  

1. -m | --metadata = path/to/seq_metadata/files # This also sets the working directory to write files
2. -f | --fasta_name = name_fasta_file_to_submit.fa # name of fasta file you will submit to GISAID
3. -s | --submitter_name = gisaid_submitter_name
4. -t | --seq_type = sequencing_type (Oxford Nanopore GridION, Illumina MiSeq, Illumina NextSeq)
5. -r | --rerun_file = path/to/rerun_file.tsv # tsv downloaded from List_of_bad_runs_and_waste_water_samples
6. -c | --completed_file = path/to/completed_file.tsv # tsv downloaded from submissions_and_compiled_metadata
7. -o | --ongoing_file = path/to/ongoing_file.tsv # tsv downloaded from submissions_and_compiled_metadata

**Example usage (short flags):**
```{bash, "gisaid_short", echo=TRUE, eval=FALSE}
# You need to type Rscript before your Rscript to run (Rscript gisaid_ncbi_biosample_metadata_formatting.R)
Rscript gisaid_ncbi_biosample_metadata_formatting.R -m /home/mandy_waters/test_metadata -f test_fasta.fa -s mandywaters -t "Oxford Nanopore GridION" -r /home/mandy_waters/rerun_samples_11Sep2021.tsv -c /home/mandy_waters/samples_already_submitted_11Sep2021.tsv -o /home/mandy_waters/ongoing_submissions_11Sep2021.tsv

# Strings with spaces need quotes
```

**Example usage (long flags):**
```{bash, "gisaid_long", echo=TRUE, eval=FALSE}
# You need to type Rscript before your Rscript to run (Rscript gisaid_ncbi_biosample_metadata_formatting.R)
Rscript gisaid_ncbi_biosample_metadata_formatting.R --metadata /home/mandy_waters/test_metadata -f test_fasta.fa --submitter_name mandywaters --seq_type "Oxford Nanopore GridION" --rerun_file /home/mandy_waters/rerun_samples_11Sep2021.tsv --completed_file /home/mandy_waters/samples_already_submitted_11Sep2021.tsv --ongoing_file /home/mandy_waters/ongoing_submissions_11Sep2021.tsv

# Strings with spaces need quotes
```

**Output to terminal:**
Script prints the strings the user input for you to check.  If you don't provide values for all flags then the script will error out and tell you which flag(s) are missing.  The script will also error out if the sequence type name is misspelled.  A warning and file will be printed / written that lists the accession name, coverage, and project number for samples that are missing a collection date or have a collection date prior to 2020.  It will not error out, but will print this info and write it to a file for users to review. Another warning (will not error out) will be print if samples in your current batch have already been submitted or have been rerun.  Samples will be dropped if they have already been submitted to GISAID or NCBI or are in the initial round of sequencing.  It will print relevant information from the COMPLETED tab or ONGOING tab (if rejected by GenBank) in the metadata tracking excel sheet on Google Drive.

You should see something like the following regardless of short or long flag names
```{r, "r_out", echo=TRUE, eval=FALSE}
Rscript /home/mandy_waters/scripts/git_repo/seq-submission-metadata-R/gisaid_ncbi_biosample_metadata_formatting.R -m ${cur_wd}/metadata -s mandy.waters -f ${batch}_gisaid_inital_upload.fa -t "Illumina NextSeq" -r /home/mandy_waters/Downloads/rerun_samples_17Oct2021.tsv -c /home/mandy_waters/Downloads/completed_submissions_17Oct2021.tsv -o /home/mandy_waters/Downloads/ongoing_submissions_17Oct2021.tsv

INPUT FILES AND FLAGS:

metadata path:
/home/mandy_waters/batch19_submit/metadata

name of fasta file to submit to GISAID:
batch19_gisaid_inital_upload.fa

submitter name for GISAID:
mandy.waters

sequencing type:
Illumina NextSeq

name of rerun file:
/home/mandy_waters/Downloads/rerun_samples_17Oct2021.tsv

name of completed submissions file:
/home/mandy_waters/Downloads/completed_submissions_17Oct2021.tsv

name of ongoing submissions file:
/home/mandy_waters/Downloads/ongoing_submissions_17Oct2021.tsv

SURVEILLANCE:
results.tsv files DO NOT have surveillance data


BAD NOODLES:
WARNING: the samples below have been submitted under multiple projects.  If you are processing the project in 'first_run' these samples will be removed from the dataset:
 accession_id   first_run     re_run
   2101678415 COVMIN_0195 NEXSEQ_050
   2101648947  NEXSEQ_046 NEXSEQ_050
   2101649102  NEXSEQ_046 NEXSEQ_050
   2101649021  NEXSEQ_046 NEXSEQ_050
   2101649115  NEXSEQ_046 NEXSEQ_050
   2101648944  NEXSEQ_046 NEXSEQ_050
   2101648933  NEXSEQ_046 NEXSEQ_050
   2101648926  NEXSEQ_046 NEXSEQ_050
   2101649116  NEXSEQ_046 NEXSEQ_050
   2101649013  NEXSEQ_046 NEXSEQ_050
   2101649118  NEXSEQ_046 NEXSEQ_050
   2101649106  NEXSEQ_046 NEXSEQ_050
   2101649220  NEXSEQ_046 NEXSEQ_050
   2101649000  NEXSEQ_046 NEXSEQ_050
   2101648994  NEXSEQ_046 NEXSEQ_050
   2101648987  NEXSEQ_046 NEXSEQ_050

FINISHED! Please check warnings and output files

```

**Output files:**

1. GISAID formatted file: gisaid_submission_date_metadata.csv (e.g. gisaid_submission_2021-07-03_metadata.csv)
2. NCBI Biosample formatted file: ncbi_biosample_submission_date_metadata.tsv (e.g. ncbi_biosample_submission_2021-07-03_metadata.tsv)
3. File to convert current fasta names to GISAID and GenBank format (fasta_rename_accession_to_gisaid_id.csv)
4. CONDITIONAL: Missing collection date only writes this if samples are missing a collection date (e.g. samples_missing_collection_date_2021-07-03.tsv)
5. CONDITIONAL: Collection date before 2020 only writes if samples collection date is before 2020 (e.g. samples_wrong_collection_date_2021-09-11.tsv)
6. CONDITIONAL: samples that have been rerun are listed if the accessions in your batch match accessions listed in the rerun file (e.g. samples_rerun_2021-09-11.tsv)
7. CONDITIONAL: samples that have already been submitted under a previous project (e.g. samples_already_submitted_2021-09-11.tsv)

I use the following awk command to replace the accession IDs in the fasta with GISAID accepted names.  These samples have been run through all the steps through indel finder.  After this the samples are run through VADR.
```{bash, "remplace_gisaid_ids", echo=TRUE, eval=FALSE}
# Replace the accession IDs with GISAID IDs
# Takes in file output #3 from gisaid_ncbi_biosample_metadata_formatting.R
awk -F, 'FNR==NR {f2[$1]=$2;next} $2 in f2 {$2=f2[$2]}1' /path/to/fasta_rename_accession_to_gisaid_id.csv FS='>' OFS='>' input_fasta_post_indel_finder.fa > output_fasta_post_indel_finder_gisaid_format.fa
```

####Error example
This is an example of an error message that's missing the -r option
```{bash, "r_out2", echo=TRUE, eval=FALSE}
Rscript gisaid_ncbi_biosample_metadata_formatting.R -m /home/mandy_waters/test_metadata -f test_fasta.fa -s mandywaters -t "Oxford Nanopore GridION" -c /home/mandy_waters/samples_already_submitted_11Sep2021.tsv


metadata path:
/home/mandy_waters

name of fasta file to submit to GISAID:
test.fa

submitter name for GISAID:
mandy.waters

sequencing type:
Oxford Nanopore GridION

name of rerun file:
NA

name of completed submissions file:
/home/mandy_waters/samples_already_submitted_11Sep2021.tsv


ERROR: you did not specify all varialbles. Please supply all required options. Type Rscript gisaid_ncbi_biosample_metadata_formatting.R -h for list of required options

Error in is.data.frame(x) : object 'ncbi_biosample_metadata' not found
Calls: write.table -> is.data.frame
Execution halted
```

###NCBI SRA {.tabset .tabset-fade .tabset-pills}
####Useage
This script takes in user specified information and file sent from NCBI biosample submisssion (BioSampleObjects.txt) to create a metadata file with format needed to submit reads to SRA. You can run this R script directly from the command line.  You will need to supply 2 pieces of information as options / flags when running the script. NOTE: this writes out the file names based on expected patterns for each type of sequecing and assumes that all files are gzipped.  I use the script get_cat_gzip_fastq.sh, written by Mike Martin with a minor adjustments to include gzipping and write all files to the same directory for concatting Oxford Nanopore reads into a single fastq.gz.  The version I use is in the same repo as these R scripts.

**Script name:** ncbi_sra_metadata_formatting.R  
**Input file in path:** BioSampleObjects.txt (file attached from NCBI BioSample email confirming samples are published)
**Options / flags:** required to fill in all.  The code will check spelling on the sequencing information and will error out if something is misspelled (options in #2 below)  

1. -p | --path = path/to/seq_metadata/files # This also sets the working directory to write files
2. -t | --seq_type = sequencing_type (Oxford Nanopore GridION, Illumina MiSeq, Illumina NextSeq)

**Example usage (short flags):**
```{bash, "sra_short", echo=TRUE, eval=FALSE}
# You need to type Rscript  before your Rscript to run (Rscript ncbi_sra_metadata_formatting.R)
Rscript ncbi_sra_metadata_formatting.R -p /home/mandy_waters/metadata -t "Oxford Nanopore GridION"

# Strings with spaces need quotes
```

**Example usage (long flags):**
```{bash, "sra_long", echo=TRUE, eval=FALSE}
# You need to type Rscript  before your Rscript to run (Rscript ncbi_sra_metadata_formatting.R)
Rscript ncbi_sra_metadata_formatting.R --path /home/mandy_waters/test_metadata --seq_type "Oxford Nanopore GridION"

# Strings with spaces need quotes
```

**Output to terminal:**
Script prints the strings the user input for you to check.  If you don't privde values for all flags then the script will error out and tell you which flag(s) are missing.  The script will also error out if the sequence type name is misspelled

You should see something like the following regardless of short or long flag names
```{r, "sra_out", echo=TRUE, eval=FALSE}
Rscript ncbi_sra_metadata_formatting.R -p /home/mandy_waters/metadata -t "Oxford Nanopore GridION"

working directory:
/home/mandy_waters/metadata

sequencing type:
Oxford Nanopore GridION
```

**Output files:**

1. SRA formatted file: ncbi_sra_submission_date_metadata.tsv (e.g. ncbi_sra_submission_2021-07-24_metadata.tsv)

Then I use the following code to get a list of the samples for which we want to upload the reads and then delete the rest.  We downloaded all the samples, but will only submit those with >50% coverage.  

```{bash, "removing_fasta", echo=TRUE, eval=FALSE}
# Get list of files to keep
cut -f 14 ncbi_sra_submission_2021-07-17_metadata.tsv | sed 's/"//g' | grep '^CO' > fastqs_to_keep.tsv

# Delete the files we don't want to keep. Run this in the directory were your gzipped fastas are located. You may need to change the path to fastqs_to_keep.tsv. To see how this runs you can always comment out rm "$i" line first to check which fastas will be deleted. Then add it back in to delete files you won't be uploading.
for i in *fastq.gz; do
    if ! grep -qxFe "$i" ../../fastqs_to_keep.tsv; then
    	echo "Deleting: $i"
        rm "$i"
    fi
done
```

####Error example
This is an example of an error message - missing the -t option
```{bash, "sra_out2", echo=TRUE, eval=FALSE}
Rscript ncbi_sra_metadata_formatting.R -m /home/mandy_waters/metadata  -t "Oxford Nanopore GRIDION"

working directory:
/home/mandy_waters/metadata

sequencing type:
Oxford Nanopore GRIDION



ERROR: check the spelling of your sequence type input: sequence type should match 'Illumina MiSeq', 'Illumina NextSeq', or 'Oxford Nanopore GridION'

Error in is.data.frame(x) : object 'biosample_metadata' not found
Calls: write.table -> is.data.frame
Execution halted
```

###NCBI Genbank {.tabset .tabset-fade .tabset-pills}
####Useage
This script takes in user specified information, downloaded file from GISAID (gisaid_hcov-19_\*.tsv), and downloaded file from NCBI SRA (metadata-.\*-processed-ok.tsv). You can run this R script directly from the command line.  You will need to supply 2 pieces of information as options / flags when running the script.

**Script name:** ncbi_genbank_metadata_formatting.R  
**Input files in path:** metadata-\*-processed-ok.tsv (downloaded from NCBI submission page - click 'Download metadata file with SRA accessions' and moved to directory you put in the -p option) and gisaid_hcov-19_\*.tsv (Download the seqeunce_metadata file from GISAID search and move to directory you put in the -p option)  
**Options / flags:** required to fill in all.  The code will check spelling on the sequencing information and will error out if something is misspelled (options in #2 below)  

1. -p | --path = path/to/seq_metadata/files # This also sets the working directory to write files
2. -t | --seq_type = sequencing_type (Oxford Nanopore GridION, Illumina MiSeq, Illumina NextSeq)

**Example usage (short flags):**
```{bash, "genbank_short", echo=TRUE, eval=FALSE}
# You need to type Rscript  before your Rscript to run (Rscript ncbi_sra_metadata_formatting.R)
Rscript ncbi_genbank_metadata_formatting.R -p /home/mandy_waters/metadata -t "Oxford Nanopore GridION"

# Strings with spaces need quotes
```

**Example usage (long flags):**
```{bash, "genbank_long", echo=TRUE, eval=FALSE}
# You need to type Rscript  before your Rscript to run (Rscript ncbi_sra_metadata_formatting.R)
Rscript ncbi_genbank_metadata_formatting.R --path /home/mandy_waters/test_metadata --seq_type "Oxford Nanopore GridION"

# Strings with spaces need quotes
```

**Output to terminal:**
Script prints the strings the user input for you to check.  If you don't privde values for all flags then the script will error out and tell you which flag(s) are missing.  The script will also error out if the sequence type name is misspelled

You should see something like the following regardless of short or long flag names
```{r, "genbank_out", echo=TRUE, eval=FALSE}
Rscript ncbi_genbank_metadata_formatting.R -p /home/mandy_waters/metadata -t "Oxford Nanopore GridION"

working directory:
/home/mandy_waters/metadata

sequencing type:
Oxford Nanopore GridION
```

**Output files:**

1. NCBI Genbank formatted file: ncbi_genbank_submission_date_metadata.tsv (e.g. ncbi_genbank_submission_2021-07-24_metadata.tsv)

####Error example
This is an example of an error message - missing the -t option
```{bash, "genbank_out2", echo=TRUE, eval=FALSE}
Rscript ncbi_genbank_metadata_formatting.R -m /home/mandy_waters/metadata  -t "Oxford Nanopore GRIDION"

working directory:
/home/mandy_waters/metadata

sequencing type:
Oxford Nanopore GRIDION



ERROR: check the spelling of your sequence type input: sequence type should match 'Illumina MiSeq', 'Illumina NextSeq', or 'Oxford Nanopore GridION'

Error in is.data.frame(x) : object 'biosample_metadata' not found
Calls: write.table -> is.data.frame
Execution halted
```

###Combined data - submission completed tab {.tabset .tabset-fade .tabset-pills}
####Useage
This script takes in user specified information, downloaded file from GISAID (gisaid_hcov-19_\*.tsv), genbank donwloaded file (seqids.txt or accessions.txt) and downloaded file from NCBI SRA (metadata-.\*-processed-ok.tsv), and results\*.tsv to pull the seq_run information. This code will output a file with combined metadata that you will add to the 'COMPLETED_submssions' tab and conditionally a file if GenBank rejected some of your samples that you will add to the 'ONGOING_submissions' tab.  You can run this R script directly from the command line.  You will need to supply 2 pieces of information as options / flags when running the script.

**Script name:** submission_completed_metadata.R  
**Input files in path:** metadata-\*-processed-ok.tsv, seqids.txt (sometimes GenBank sends this file as accessions.txt - code will take either file name), and gisaid_hcov-19_\*.tsv  
**Options / flags:** required to fill in all.  The code will check if the user specificed both options or it will error out.

1. -p | --path = path/to/seq_metadata/files # This also sets the working directory to write files
2. -s | --submitter_name = submitter name

**Example usage (short flags):**
```{bash, "completed_short", echo=TRUE, eval=FALSE}
# You need to type Rscript  before your Rscript to run (Rscript ncbi_sra_metadata_formatting.R)
Rscript submission_completed_metadata.R -p /home/mandy_waters/metadata -s mandy.waters

# Strings with spaces need quotes
```

**Example usage (long flags):**
```{bash, "completed_long", echo=TRUE, eval=FALSE}
# You need to type Rscript  before your Rscript to run (Rscript ncbi_sra_metadata_formatting.R)
Rscript submission_completed_metadata.R --path /home/mandy_waters/test_metadata --submitter_name mandy.waters

# Strings with spaces need quotes
```

**Output to terminal:**
Script prints the strings the user input for you to check.  If you don't privde values for all flags then the script will error out and tell you which flag(s) are missing.  

You should see something like the following regardless of short or long flag names
```{r, "completed_out", echo=TRUE, eval=FALSE}
Rscript submission_completed_metadata.R -p /home/mandy_waters/metadata -s mandy.waters

working directory:
/home/mandy_waters/metadata

submitter name:
mandy.waters
```

**Output files:**

1. Compiled metadata for samples accepted by all submission sites to be tracked in 'COMPLETED_submissions' tab: COMPLETED_date_metadata.tsv (e.g. COMPLETED_2021-10-10_metadata.tsv)
2. CONDITIONAL: compiled metadata for samples rejected by GenBank to be tracked in 'ONGOING_submissions' tab: ONGOING_genbank_missing_date_metadata.tsv (e.g. ONGOING_genbank_missing_2021-10-10_metadata.tsv)

####Error example
This is an example of an error message that's missing the -s option
```{bash, "coomplete_error", echo=TRUE, eval=FALSE}
Rscript submission_completed_metadata.R -p /home/mandy_waters/metadata

working directory:
/home/mandy_waters/test_metadata

submitter name:
NA

ERROR: you did not specify all varialbles. Please supply all values

Error in is.data.frame(x) : object 'ncbi_biosample_metadata' not found
Calls: write.table -> is.data.frame
Execution halted
```
